clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1442,13 +1527,46 @@ class SyclKernelBodyCreator
     // TODO: Creates init/finalize sequence and inits special sycl obj
   }
 
+  void CreateExprForStructOrScalar(FieldDecl *FD) {
+    ParmVarDecl *KernelParameter =
+        DeclCreator.getParamVarDeclsForCurrentField()[0];
+    InitializedEntity Entity =
+        InitializedEntity::InitializeMember(FD, &VarEntity);
+    QualType ParamType = KernelParameter->getOriginalType();
+    Expr *DRE = SemaRef.BuildDeclRefExpr(KernelParameter, ParamType, VK_LValue,
+                                         SourceLocation());
+    InitializationKind InitKind =
+        InitializationKind::CreateCopy(SourceLocation(), SourceLocation());
+    InitializationSequence InitSeq(SemaRef, Entity, InitKind, DRE);
+
+    ExprResult MemberInit = InitSeq.Perform(SemaRef, Entity, InitKind, DRE);
+    InitExprs.push_back(MemberInit.get());
+  }
 
   void handleStructType(const FieldDecl *FD, QualType Ty) final {
-    // TODO: a bunch of work doing inits, note this has a little more than
-    // scalar.
+    CreateExprForStructOrScalar(const_cast<FieldDecl *>(FD));
   }
+
   void handleScalarType(const FieldDecl *FD, QualType Ty) final {
-    // TODO: a bunch of work doing inits.
+    CreateExprForStructOrScalar(const_cast<FieldDecl *>(FD));
+  }
+
+  void enterStruct(const CXXRecordDecl *, const FieldDecl *FD) final {
+    CurBase = BuildMemberExpr(CurBase, const_cast<FieldDecl *>(FD));
```

Comment:
These are structs that are 'fields' up the way.  I guess I don't totally understand the 'base' naming.

Reply:
Also, why store 'CurBase' separate, rather than just accessing it via Bases.back?

Reply:
`Base` here means not class "base" in inheritance. It is "base" for member expressions (https://clang.llvm.org/doxygen/classclang_1_1MemberExpr.html), i.e. some value to which operator `.` or `->` is applied, I will re-think naming.